template/{% if use_git and dev_platform == 'GitHub' %}.github{% endif %}/workflows/release.yml

```
@@ -1,53 +1,74 @@
-name: CD
+name: Publish to PyPI
 
 on:
-  push:
-    tags:
-    - 'v?[0-9]+.[0-9]+.[0-9]+'
-    - 'v?[0-9]+.[0-9]+.[0-9]+(a|b|rc|post|dev)[0-9]+'
-
-jobs:
-  prerequisites:
-    uses: ./.github/workflows/test.yml
-
   release:
-    needs: [prerequisites]
-    strategy:
-      matrix:
-        os: [ubuntu-latest]
-        python-version: ["3.12"]
-    runs-on: ${{ matrix.os }}
-    permissions:
-      # Write permissions are needed to create OIDC tokens.
-      id-token: write
-      # Write permissions are needed to make GitHub releases.
-      contents: write
+    types: [published]
 
+jobs:
+  # Setup build separate from publish for added security
+  # See https://github.com/pypa/gh-action-pypi-publish/issues/217#issuecomment-1965727093
+  build:
+    runs-on: ubuntu-latest
+    # This ensures that the publish action only runs in the main repository
+    # rather than forks
+    # Environment is encouraged for increased security
+    environment: build
     steps:
-    - uses: actions/checkout@v4
-
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
-      with:
-        python-version: ${{ matrix.python-version }}
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          # This fetch element is only important if you are use SCM based
+          # versioning (that looks at git tags to gather the version)
+          fetch-depth: 100
+            # Need the tags so that setuptools-scm can form a valid version number
+      - name: Fetch git tags
+        run: git fetch origin 'refs/tags/*:refs/tags/*'
```

This can be simplified with the `fetch-tags:` input, which fetches all tags without needing to fetch up to a higher depth.

```suggestion
      - name: Checkout
        uses: actions/checkout@v4
        with:
          # This fetch element is only important if you are use SCM based
          # versioning (that looks at git tags to gather the version).
          # setuptools-scm needs tags to form a valid version number
          fetch-tags: true
```

I have to note that this isn't working with `ubuntu-24.04` images, though (https://github.com/actions/checkout/issues/2041), but it looks like it will be resolved soon, well before `ubuntu-latest` starts pointing to it.

---

ahhhh ok perhaps we hold off if this is bleeding edge and not vetted across platforms. As we do want our users to have a happy experience creating their first package! It's more important that this template is easy to use and understand!

---

Oh, it's not an entirely new feature (it calls a variant of `git fetch` underneath), actually. The fix will be rolled out soon enough, reducing one step here, so I'll suggest we keep these, but I'm fine with waiting until we know it's working everywhere!

---

cool! Thanks, @agriyakhetarpal. What do you think about us moving this to an issue and applying when the bugs have been ironed out on Linux/Ubuntu? All of your comments are awesome BTW - THANK YOU for taking the time to review. I'm really excited to teach with this template in the spring!

---

Ok so I believe that you both think we should just implement this. now as the bug will be fixed soon? @Midnighter i'll go ahead and accept this change and we can move forward.

This is a case where i don't f ully understand the implications of the bug (ie how many people will run into that issue and when it will be fixed) but let's move forward with the preferable approach!